internal/gengapic/genrest.go

```diff
@@ -56,62 +80,515 @@ func (g *generator) genRESTMethods(serv *descriptor.ServiceDescriptorProto, serv
+	pathParams := map[string]*descriptor.FieldDescriptorProto{}
+	info, err := getHTTPInfo(m)
+	if info == nil || err != nil {
+		return pathParams
+	}
+
+	// Match using the curly braces but don't include them in the grouping.
+	re := regexp.MustCompile("{([^}]+)}")
+	for _, p := range re.FindAllStringSubmatch(info.url, -1) {
+		// In the returned slice, the zeroth element is the full regex match,
+		// and the subsequent elements are the sub group matches.
+		// See the docs for FindStringSubmatch for further details.
+		param := p[1]
+		field := g.lookupField(m.GetInputType(), param)
+		if field == nil {
+			continue
+		}
+		pathParams[param] = field
+	}
 
-	// clientName := camelToSnake(serv.GetName())
-	// clientName = strings.Replace(clientName, "_", " ", -1)
+	return pathParams
+}
 
-	// NOTE(dovs): enable this function only when we can verify that it works.
-	// If we leave it commented out, we can change it, and we don't have to
-	// guarantee correctness or stability.
-	//
-	// p("// New%sRESTClient creates a new %s rest client.", servName, clientName)
-	// g.serviceDoc(serv)
-	// p("func New%[1]sRESTClient(ctx context.Context, opts ...option.ClientOption) (*%[1]sClient, error) {", servName)
-	// p("    c := &%s{", servName)
-	// p("    }")
-	// p("")
-	// p("    return &%sClient{internal%[1]sClient: c, CallOptions: default%[1]sCallOptions()}, nil", servName)
-	// p("}")
+func (g *generator) queryParams(m *descriptor.MethodDescriptorProto) map[string]*descriptor.FieldDescriptorProto {
+	queryParams := map[string]*descriptor.FieldDescriptorProto{}
+	info, err := getHTTPInfo(m)
+	if info == nil || err != nil {
+		return queryParams
+	}
+	pathParams := g.pathParams(m)
+	// Minor hack: we want to make sure that the body parameter, if one exists,
+	// is NOT a query parameter.
+	pathParams[info.body] = nil
+	msg := g.descInfo.Type[m.GetInputType()].(*descriptor.DescriptorProto)
+	if info.body != "" && info.body != "*" {
+		field := g.lookupField(m.GetInputType(), info.body)
+		msg = g.descInfo.Type[field.GetTypeName()].(*descriptor.DescriptorProto)
+	}
+
+	// If, and only if, a field is a path parameter, it's not a query parameter.
+	for _, f := range msg.Field {
+		if _, ok := pathParams[f.GetName()]; !ok {
+			queryParams[f.GetName()] = f
+		}
+	}
 
+	return queryParams
+}
+
+func (g *generator) generateQueryString(m *descriptor.MethodDescriptorProto) {
+	p := g.printf
+	queryParams := g.queryParams(m)
+	if queryParams == nil {
+		return
+	}
+
+	// We want to iterate over fields in a deterministic order
+	// to prevent spurious deltas when regenerating gapics.
+	fields := make([]string, 0, len(queryParams))
+	for p, _ := range queryParams {
+		fields = append(fields, p)
+	}
+	sort.Strings(fields)
+	info, _ := getHTTPInfo(m)
+
+	bodyStr := "req"
+	if info.body != "" && info.body != "*" {
+		bodyStr = "body"
+		p("body := req.Get%s()", snakeToCamel(info.body))
```

Comment:
Okay, so to clarify, using pseudopython set notation:
pathParams = set(fields appearing in the url path)
bodyParams = set(

The companion showcase PR is here: https://github.com/googleapis/gapic-showcase/pull/751

Reply:
Correct, with the following minor caveats:
- bodyParams may overlap pathParams (eg in case of "*"). The server will prioritize the path param values
- query params: the set of fields you specified above, but each one subject to the restriction that it is encoded iff 
   ```
  ( (field.Value()!=field.DefaultValue())    ||      (field.isOptional() && field.isSet())   )
  ```

See https://cloud.google.com/endpoints/docs/grpc-service-config/reference/rpc/google.api#rules-for-http-mapping and my [comment](https://github.com/googleapis/gapic-showcase/pull/751/files#r652210323) on the Showcase PR

Reply:
But what's trippingme up is that your `queryParams()` above is implementing this logic, so it's just this `generateQueryString()` that is confusing.

Reply:
BTW: I _think_ (please verify) that the condition I quote above for query params can be simplified in Go by seeing whether the proto message field is null or not (ie not using accessor methods, which try to hide that implementation detail). But it might arguably be more elegant to use accessors and have the explicit check.

Reply:
Okay, made the fix. Thanks for pointing this out, I don't know how I misread the spec so badly.